pkg/controller/usersignup/predicate.go

```diff
@@ -44,6 +46,11 @@ func (p UserSignupChangedPredicate) LabelChanged(e event.UpdateEvent, labelName
 	return e.MetaOld.GetLabels()[labelName] != e.MetaNew.GetLabels()[labelName]
 }
 
+func (p UserSignupChangedPredicate) Delete(e event.DeleteEvent) bool {
+	// will reconcile to remove the finalizers
+	return len(e.Meta.GetFinalizers()) > 0
```

Comment:
> If I understand this correctly you want to prevent decrementing the counter multiple times while deleting the same usersignup by using this predicate. Maybe I'm wrong but I feel a bit uncomfortable using predicates for that...
> 

To clarify, I set a finalizer on the UserSignup resource to make sure that we can process the resource when it's deleted (ie, decrement the counter). 


> What if we do something like this instead:
> 
> * Set the activation annotation only when we successfully created a MUR (account is truly active). No MUR == no annotation.
> * Add the finalizer if it it's not present

ok, I can do that

> * If the UserSignup is being deleted then:
> 
> ```
> if (usersignup.annotation is present) {
>     // Delete the annotation
>     // Decrement the counter (only after successfully deleting the annotation)
>     // Delete the finalizer
>     return
> } else {
>     // the signup is not active. Don't need to do anything. Delete the finalizer and return
> }
> ```

what's the benefit of deleting the annotation? Isn't it the finalizer's presence that tells if the controller should do some cleanup while the resource is deleted?

Reply:
We can't fully control the reconcile calls. So theoretically another reconcile could happen while we are decrementing the counter. So, the non-existing annotation would tell the reconcile not to decrement the counter one more time.
I'm trying to make it level/state driven (vs. edge/even driven) as much as possible.

Reply:
in that case, why can't we just do something like this:

```
if (usersignup.finalizer is present) {
    // Delete the finalizer
    // Decrement the counter (only after successfully deleting the annotation)
    return
} else {
    // the signup is not active. Don't need to do anything (ie, we don't even need this 'else' branch)
}
```

Reply:
Yeah... maybe...
Just got used to deleting the finalizers when the job is done.. :)
@MatousJobanek thoughts?

Reply:
But we need to check the annotation anyway. We do not want to decrement the counter when deleting an inactive user (no annotation present).

Reply:
what do you mean by inactive user and annotation not present? 
Firstly, the controller will set the annotation on all existing resources during the start-up. Granted, there could be an edge-case (or race condition) where a usersignup is deleted before the annotation was set
Secondly, this is already covered in https://github.com/xcoulon/host-operator/blob/IssueCRT988_metric_user_reactivations/pkg/controller/usersignup/usersignup_controller.go#L166 since we need to decrement the counter for a specific number of activations. And for that, we need to read the annotation ;)

Reply:
You missed my comment above:

> Set the activation annotation only when we successfully created a MUR (account is truly active). No MUR == no annotation.

We don't want to count inactive accounts (when pending verification for example).

Reply:
I'm not a fan of the predicate either and - TBH - I don't really see the purpose of the predicate. 
If we want to do something like that - decrement the counter, I would rather put the logic into `usersignup_cleanup_controller.go`. I know that we can still delete the usersignups via gdpr-delete, but how many of them do we have? Almost nothing

Reply:
the logic to decrement the counter when the user was deleted has been removed in e397612